src/common/device_helpers.cuh

```diff
@@ -866,25 +870,31 @@ class AllReducer {
    * group.
    *
    * \param device_ordinals The device ordinals.
+   * \param opg Whether to initialize AllReducer with a single process per GPU
    */
 
   void Init(const std::vector<int> &device_ordinals) {
 #ifdef XGBOOST_USE_NCCL
     /** \brief this >monitor . init. */
     this->device_ordinals = device_ordinals;
     comms.resize(device_ordinals.size());
-    dh::safe_nccl(ncclCommInitAll(comms.data(),
-                                  static_cast<int>(device_ordinals.size()),
-                                  device_ordinals.data()));
+    auto id = GetUniqueId();
+
+    for (size_t i = 0; i < device_ordinals.size(); i++) {
+      int rank = rabit::GetRank() + device_ordinals[i];
```

Comment:
Note that the computation of this rank is not general.

Reply:
Maybe change this variable to be called nccl_rank to differentiate it from the rabit rank.

I think this should be: 
```
int nccl_rank = rabit::GetRank() * device_ordinals.size() + device_ordinals[i];
```
For example if each rabit rank has more than one GPU, your calculation above will have the nccl ranks overlapping.

Note that this calculation is only correct so long as each rabit rank has the same number of GPUs.

Reply:
Thanks for the catch. I've corrected this in the next commit.